File: php/src/Tls/NamedGroups.php
<?php

declare(strict_types=1);

namespace PQC\Tls;

/**
 * TLS 1.3 Named Groups for PQC key exchange.
 * Based on draft-ietf-tls-mlkem and hybrid proposals.
 */
final class NamedGroups
{
    // ML-KEM standalone groups
    public const MLKEM512 = 0x0200;
    public const MLKEM768 = 0x0201;
    public const MLKEM1024 = 0x0202;

    // Hybrid groups (X25519 + ML-KEM)
    public const X25519_MLKEM768 = 0x6399;  // Standardized in RFC 9562
    public const X25519_MLKEM512 = 0x639A;

    // Hybrid groups (SecP256r1 + ML-KEM)
    public const SECP256R1_MLKEM768 = 0x639B;

    // Classical groups (for reference)
    public const X25519 = 0x001D;
    public const X448 = 0x001E;
    public const SECP256R1 = 0x0017;
    public const SECP384R1 = 0x0018;
    public const SECP521R1 = 0x0019;

    /**
     * Get all PQC-related named groups.
     *
     * @return array<int, array{name: string, type: string, pqc: bool}>
     */
    public static function all(): array
    {
        return [
            self::MLKEM512 => [
                'name' => 'MLKEM512',
                'type' => 'kem',
                'pqc' => true,
                'hybrid' => false,
                'nistLevel' => 1,
            ],
            self::MLKEM768 => [
                'name' => 'MLKEM768',
                'type' => 'kem',
                'pqc' => true,
                'hybrid' => false,
                'nistLevel' => 3,
            ],
            self::MLKEM1024 => [
                'name' => 'MLKEM1024',
                'type' => 'kem',
                'pqc' => true,
                'hybrid' => false,
                'nistLevel' => 5,
            ],
            self::X25519_MLKEM768 => [
                'name' => 'X25519MLKEM768',
                'type' => 'hybrid_kem',
                'pqc' => true,
                'hybrid' => true,
                'nistLevel' => 3,
                'classical' => 'x25519',
            ],
            self::X25519_MLKEM512 => [
                'name' => 'X25519MLKEM512',
                'type' => 'hybrid_kem',
                'pqc' => true,
                'hybrid' => true,
                'nistLevel' => 1,
                'classical' => 'x25519',
            ],
            self::SECP256R1_MLKEM768 => [
                'name' => 'SecP256r1MLKEM768',
                'type' => 'hybrid_kem',
                'pqc' => true,
                'hybrid' => true,
                'nistLevel' => 3,
                'classical' => 'p256',
            ],
        ];
    }

    /**
     * Get info for a specific group code.
     */
    public static function get(int $code): ?array
    {
        $all = self::all();
        return $all[$code] ?? null;
    }

    /**
     * Check if a named group is PQC.
     */
    public static function isPqc(int $code): bool
    {
        $info = self::get($code);
        return $info !== null && $info['pqc'];
    }

    /**
     * Check if a named group is hybrid.
     */
    public static function isHybrid(int $code): bool
    {
        $info = self::get($code);
        return $info !== null && ($info['hybrid'] ?? false);
    }

    /**
     * Get TLS extension bytes for supported_groups.
     *
     * @param array<int> $groups Array of group codes
     * @return string Wire-format bytes
     */
    public static function encodeSupported(array $groups): string
    {
        $body = '';
        foreach ($groups as $code) {
            $body .= pack('n', $code);
        }
        return pack('n', strlen($body)) . $body;
    }

    /**
     * Decode supported_groups extension.
     *
     * @return array<int> Group codes
     */
    public static function decodeSupported(string $bytes): array
    {
        $len = unpack('n', substr($bytes, 0, 2))[1];
        $groups = [];
        for ($i = 0; $i < $len; $i += 2) {
            $groups[] = unpack('n', substr($bytes, 2 + $i, 2))[1];
        }
        return $groups;
    }
}

File: dotnet/src/PqcStandards/Tls/NamedGroups.cs
namespace PqcStandards.Tls;

/// <summary>TLS 1.3 named groups for PQC key exchange (RFC 9180 / draft-ietf-tls-hybrid-design).</summary>
public static class NamedGroups
{
    // Traditional
    public const ushort X25519 = 0x001D;
    public const ushort Secp256r1 = 0x0017;
    public const ushort Secp384r1 = 0x0018;
    public const ushort Secp521r1 = 0x0019;

    // PQC standalone
    public const ushort MlKem512 = 0x0200;
    public const ushort MlKem768 = 0x0201;
    public const ushort MlKem1024 = 0x0202;

    // Hybrid groups
    public const ushort X25519MlKem768 = 0x6399;
    public const ushort Secp256r1MlKem768 = 0x639A;
    public const ushort Secp384r1MlKem1024 = 0x639B;

    /// <summary>Get the human-readable name for a named group.</summary>
    public static string GetName(ushort group) => group switch
    {
        X25519 => "x25519",
        Secp256r1 => "secp256r1",
        Secp384r1 => "secp384r1",
        Secp521r1 => "secp521r1",
        MlKem512 => "ML-KEM-512",
        MlKem768 => "ML-KEM-768",
        MlKem1024 => "ML-KEM-1024",
        X25519MlKem768 => "X25519MLKEM768",
        Secp256r1MlKem768 => "SecP256r1MLKEM768",
        Secp384r1MlKem1024 => "SecP384r1MLKEM1024",
        _ => $"Unknown(0x{group:X4})"
    };

    /// <summary>Check if a group is PQC or hybrid.</summary>
    public static bool IsPqc(ushort group) => group >= 0x0200;

    /// <summary>Check if a group is a hybrid group.</summary>
    public static bool IsHybrid(ushort group) => group >= 0x6399 && group <= 0x639B;

    /// <summary>All supported PQC and hybrid groups.</summary>
    public static readonly ushort[] AllPqcGroups = [MlKem512, MlKem768, MlKem1024, X25519MlKem768, Secp256r1MlKem768, Secp384r1MlKem1024];
}
